### PR TITLE
feat(vm-route-forge): add ebpf route watcher

### PR DESCRIPTION
**Pull request by yaroslavborbat**
## Description
Add the ebpf route watcher.

Insert ebpf hooks to run in kernel space before executing the fib_table_insert and fib_table_delete system calls and generate a route change event and save it on the FIFO map.

Our watcher in the user space will read and delete events from the bpf map. Thanks to these events, he will be able to identify the virtual machines that need to be queued for route synchronization.

## Why do we need it, and what problem does it solve?

The [netlink.RouteSubscribe](https://pkg.go.dev/github.com/vishvananda/netlink#RouteSubscribe) consumes too much processor resources. We need an alternative way to receive route change events.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/vm-route-forge/cmd/vm-route-forge/app/options/options.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ func (o *Options) Flags(fs *pflag.FlagSet) {
 	fs.StringVarP(&o.NodeName, flagNodeName, flagNodeNameShort, os.Getenv(NodeNameEnv), "The name of the node.")
 	fs.StringVarP(&o.RouteTableID, flagTableId, flagTableIdShort, os.Getenv(RouteTableIDEnv), "The id of the table.")
 	fs.IntVarP(&o.Verbosity, flagVerbosity, flagVerbosityShort, getDefaultVerbosity(), "Verbosity of output.")
-	fs.StringVar(&o.KindRouteWatcher, flagKindRouteWatcher, getEnvWithDefault(KindRouteWatcherEnv, string(route.TickerKind)), "Kind of route watcher.")
+	fs.StringVar(&o.KindRouteWatcher, flagKindRouteWatcher, getEnvWithDefault(KindRouteWatcherEnv, string(route.EbpfKind)), "Kind of route watcher.")
```
Comment (LopatinDmitr) [inline images/vm-route-forge/cmd/vm-route-forge/app/options/options.go:79]: Do we want to set eBPF by default?

File: images/vm-route-forge/internal/netlinkmanager/manager.go [hunk 2]
```diff
@@ -17,31 +17,32 @@ limitations under the License.
...
 	"vm-route-forge/internal/netutil"
 )
 
+type Action uint32
+
+const (
+	ActionAdd Action = iota
+	ActionDelete
+)
+
```
Comment (diafour) [inline images/vm-route-forge/internal/netlinkmanager/manager.go:45]: Is it used here?

File: images/vm-route-forge/internal/controller/route/ebpf.go [hunk 3]
```diff
@@ -18,23 +18,252 @@ package route
...
+	log := w.log.WithValues(
+		"inHostVMIP", vmIP.String(),
+		"virtualMachine", key)
+	log.Info("Started processing route")
+
+	switch event.Action {
+	case ActionAdd:
+		addrs, found := w.cache.GetAddresses(key)
+		if !found {
+			log.Info("The route was added, but there is no addresses in the cache. Add the VM to the queue.")
```
Comment (diafour) [inline images/vm-route-forge/internal/controller/route/ebpf.go:227]:
```suggestion
			log.Info("The route was added, but there are no addresses in the cache. Add the VM to the queue.")
```

File: images/vm-route-forge/Taskfile.yaml [hunk 4]
```diff
@@ -2,7 +2,34 @@
...
+      - sh: if [ -z $DLV_IMAGE ]; then exit 1 ; fi
+        msg: "export env {{ .DLV_IMAGE_ENV }}"
+    cmd: cd ../../ && docker build -f ./images/vm-route-forge/dlv.Dockerfile -t "${DLV_IMAGE}" .
+
+  dlv:push:
+    desc: "Push image vm-route-forge with dlv"
+    preconditions:
+      - sh: if [ -z $DLV_IMAGE ]; then exit 1 ; fi
+        msg: "export env {{ .DLV_IMAGE_ENV }}"
+    cmd: cd ../../ && docker build -f ./images/vm-route-forge/dlv.Dockerfile -t "${DLV_IMAGE}" .
```
Comment (diafour) [inline images/vm-route-forge/Taskfile.yaml:29]: 1. These tasks are same: cmd is docker build in both cases.
2. Why define DLV_IMAGE_ENV if `$DLV_IMAGE` is used ins script?
3. `msg: "export env {{ .DLV_IMAGE_ENV }}"` -> `msg: "Set DLV_IMAGE to dlv image name, e.g. <add example image name>`
4. Why there is no default dlv image?

File: images/vm-route-forge/bpf/route_watcher.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    ret = bpf_probe_read_kernel(&evt.src, sizeof(evt.src), &cfg->fc_prefsrc);
+    if (ret < 0) {
+        static const char msg[] = "Failed to read src: %d";
+        bpf_trace_printk(msg, sizeof(msg), ret);
+        return ret;
+    }
+
+    return bpf_map_push_elem(&route_events_map, &evt, BPF_ANY);
```
Comment (diafour) [inline images/vm-route-forge/bpf/route_watcher.c:69]: These bpf_* methods requires comments for future us to understand what happen here.

File: images/vm-route-forge/internal/netlinkmanager/manager.go (hunk 2 above)
Comment (yaroslavborbat) [reply images/vm-route-forge/internal/netlinkmanager/manager.go:45]: no, fixed

File: images/vm-route-forge/cmd/vm-route-forge/app/options/options.go (hunk 1 above)
Comment (yaroslavborbat) [reply images/vm-route-forge/cmd/vm-route-forge/app/options/options.go:79]: no, fixed

File: images/vm-route-forge/bpf/route_watcher.c (hunk 5 above)
Comment (yaroslavborbat) [reply images/vm-route-forge/bpf/route_watcher.c]: add comments

**Review by LopatinDmitr — DISMISSED** [review]

File: images/vm-route-forge/internal/controller/route/route_controller.go [hunk 6]
```diff
@@ -203,44 +205,41 @@ func (c *Controller) queueAdd(key string) {
...
-		keyCh, err := c.routeWatcher.Watch(ctx)
-		if err != nil {
-			errCh <- err
-			return
-		}
-		for key := range keyCh {
+		for key := range c.routeWatcher.ResultChanel() {
 			c.queueAdd(key.String())
 		}
+		watcherDone <- struct{}{}
```
Comment (yaroslavborbat) [inline images/vm-route-forge/internal/controller/route/route_controller.go:234]: close or cancel

File: images/vm-route-forge/internal/controller/route/gen.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package route
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target amd64 -type route_event ebpf ../../../bpf/route_watcher.c
```
Comment (yaroslavborbat) [inline images/vm-route-forge/internal/controller/route/gen.go:19]: add description about no export gen structs

**Review by diafour — APPROVED** [review]

**Pull request merged**